### ISSUE TITLE
[air] [rfc] Re-introduce session.checkpoint_dir() API

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#30070]
Creating directory checkpoints in Ray AIR is currently clunky:

```
temp_dir = tempfile.mkdtemp()
with open(os.path.join(temp_dir, "checkpoint.bin"), "w") as f:
  save_checkpoint(f)
session.report(metrics, checkpoint=Checkpoint.from_directory(temp_dir))
shutil.rmtree(temp_dir)
```

Saving data into a directory is a common use case. Most other libraries only support this mode (and no dict checkpoints).

From the user side, the requirement is "Just give me a dir to write data to". Creating and keeping track of the directory is overhead.

From the technical side, we need to store data into directories at some point anyway for durability (cloud storage).

Suggested options:

```
with session.checkpoint_dir() as dir:
  with open(os.path.join(temp_dir, "checkpoint.bin"), "w") as f:
      save_checkpoint(f)
  session.report(metrics, checkpoint=Checkpoint.from_directory(dir))
```

or

```
with session.checkpoint_dir(metrics) as dir:
  with open(os.path.join(temp_dir, "checkpoint.bin"), "w") as f:
      save_checkpoint(f)
```

**Comment by krfricke** [comment]
Here are a few proposals to discuss.

Base imports:

```
import os
import tempfile

from ray import tune
from ray.air import session, Checkpoint
```


# 1: Document using tempfile.TemporaryDirectory

```
def train(config):
    with tempfile.TemporaryDirectory() as tmpdir:
        with open(os.path.join(tmpdir, "checkpoint.txt"), "w") as f:
            f.write("Data\n")
        session.report({"metric": 1}, checkpoint=Checkpoint.from_directory(tmpdir))

```

Pros:
- No changes to `session.report` necessary
- No new API

Cons:
- We handle a user-generated directory
- If a user passes a _non_-temporary directory, we will move/remove it
- Not intuitively clear who manages the dir (who has to clean up) (If user does not clean up, dir will continue to exist)

# 2: `session.checkpoint_dir`, separate `session.report` call

```
def train(config):
    with session.checkpoint_dir() as checkpoint_dir:
        with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
            f.write("Data\n")
        session.report({"metric": 1}, checkpoint=Checkpoint.from_directory(checkpoint_dir))
```

Pros:
- User interacts with Ray AIR-provided checkpoint directory, does not have to create themselves
- It is implicit that Ray AIR manages the directory and cleans it up if needed

Cons:
- Weird to re-specify directory in `session.report` - what happens if the user forgets to specify?

# 3: `session.checkpoint_dir`, separate `session.report` call without checkpoint specification

```
def train(config):
    with session.checkpoint_dir() as checkpoint_dir:
        with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
            f.write("Data\n")
        session.report({"metric": 1})  # Automatically track checkpoint in this context
```

Pros:
- See 2

Cons:
- Different behavior if `session.report` is called after a checkpoint vs. not after a checkpoint
- It is not clear that a user has to call `session.report` in order to have the checkpoint tracked

# 4: `session.report_checkpoint(metrics)` with _no_ `session.report`

```
def train(config):
    with session.report_checkpoint(metrics) as checkpoint_dir:
        with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
            f.write("Data\n")
```

Pros:
- Checkpoint dir + report is explicit

Cons:
- Separate API to `session.report` (for dict checkpoints the code looks completely different)
- Feels off to specify the metrics in the context manager
- "Auto-reporting" could be unexpected by the user

**Comment by amogkam** [comment]
I would prefer not to go with 3 and 4 since these introduce yet another checkpointing api, and open up the possibility for reports and checkpoints to be called a different number of times.

**Comment by amogkam** [comment]
Actually for option 2, is it possible to do this via the Checkpoint api itself rather than the session api?

this boilerplate removal can still be useful for more general cases right?

**Comment by krfricke** [comment]
You mean something like

```
def train(config):
    with Checkpoint.from_new_directory() as checkpoint, checkpoint_dir:
        with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
            f.write("Data\n")
    session.report({"metric": 1}, checkpoint=checkpoint)
```

Or did I misunderstand?

**Comment by krfricke** [comment]
As a general comment, we will always have the possibility to report and checkpoint a different number of times, and this is probably also what we want. It's default behavior e.g. in class trainables where we specifically specify the number of steps between checkpoints (`checkpoint_frequency`)

**Comment by amogkam** [comment]
> You mean something like

yep!

**Comment by amogkam** [comment]
> As a general comment, we will always have the possibility to report and checkpoint a different number of times

Got it right. I guess I meant more like it’s explicit which metrics are associated with which checkpoint, so there’s a possibility for more user mistakes here

**Comment by krfricke** [comment]
I'm happy with that option as well. I think the main difference is that in the session API, we create the end directory itself and don't move anything, and in the Checkpoint API, we create a temp directory that we have to move. We can clean up the former automatically (e.g. if we don't end up reporting), but the latter may take up space if we don't report.

I don't think this is a heavy con though, especially because the checkpoint will be created in a temporary directory/filesystem.

Any other thoughts on this?

**Comment by xwjiang2010** [comment]
Assuming that we don't have any existing APIs, I prefer something very explicit like:
1. `session.report()` is only reporting metrics
2. `with session.checkpoint_dir(metrics) as checkpoint_dir` is only about reporting with a directory checkpoint. No separate call to `session.report` is needed. This is to avoid the confusion as Kai mentioned above.

These are my assumptions:
1. user wants to call report without checkpointing (covered by 1)
2. user wants to save checkpoint and when that happens there is always metrics associated (covered by 2)
3. user wants to also save dict. This can be achieved with 2, with something like"
```
with session.checkpoint_dir({"metrics": 1}) as checkpoint_dir:
  write_dict_to_checkpoint_dir(whatever_I_want_to_checkpoint_as_dict)  # `write_dict_to_checkpoint_dir` provided as helper function to user. 
  model.save(checkpoint_dir)
```

Other things I am still thinking about:
1. how does this work with Antoni's lazy checkpoint? What directories do we consider to be permanent? Do we want to manage train session the same as tune session? Do we want to have `rank0/checkpoint_00000` instead of just `rank0`?
2. How easy the proposed API is to work with framework specific checkpoints? Is that sufficient for at least that? To support Balaji's proposal of adding arbitrary state to TorchCheckpoint for example?
3. What does the artifact checkpoint API look like (See PRD)? Does that API have a coherent/consistent feel as this API? Will there be any confusion between the two?

**Comment by krfricke** [comment]
@xwjiang2010 just had a call about the topic.

TLDR: We 've converged on option 2. The main benefit over option 3 is that we don't want to re-introduce the implicit behavior of saving checkpoints with session.report (and not saving them without a call to `session.report`).

here are our discussion notes:

Discussion of proposals
- tempfile.TemporaryDirectory
  - If we manage checkpoint directories, they should not be created by users
  - Otherwise they implicitly (not explicitly) hand over control to us
  - This can be confusing, they could try to write to the directory later, it will fail
- session.checkpoint_dir with separate session.report call (including Checkpoint.from_directory()
  - Weird to re-specify directory again (creating Checkpoint object)  - 
  - What happens if the user forget to re-specify?
- session.checkpoint_dir, separate session.report call without checkpoint specification
  - Different behavior if session.report is called after a checkpoint vs. not after a checkpoint
  - It is not clear that a user has to call session.report in order to have the checkpoint tracked
- session.checkpoint_dir(metrics) with no session.report
  - New API (compared to current "session" API)
  - Specifying metrics in checkpoint dir creation feels a bit off
  - Two separate APIs to report metrics: Confusing to have different APIs to report dir and dict checkpoints
- Checkpoint.from_new_directory()
  - Basically just a wrapper around tempfile and Checkpoint.from_dir
  - Checkpoint directory is still not managed by AIR because unknown to session

With three supporters (plus @Yard1 as number 4 if I remember correctly) we'll go ahead with option 2. I'll prepare a PR for this shortly.

**Comment by Yard1** [comment]
I like 4, but 2 is second best for me!

**Comment by xwjiang2010** [comment]
Thanks Kai for posting the summary.

I think long term we should converge to option 4. It offers the most clear semantics and can also cover dict case reasonably well (if you are ok with a `write_dict_to_dir` utility function). #4 also means that we would change the way framework specific checkpoint is gonna look like (for better). Instead of something like:
```
model.save("my_model.h5")
checkpoint = TensorflowCheckpoint.from_h5("my_model.h5")
session.report({"my_metric": 1}, checkpoint=checkpoint)
```
, which is not clear about directory ownership
It could look like:

```
with session.checkpoint_dir(metrics) as checkpoint_dir:
  TensorflowCheckpoint.save_model(dir=checkpoint_dir, model=model, format="h5", dict={"some_object": some_object})
```

However, 2 is what can be done for 2.2.

**Comment by matthewdeng** [comment]
I actually prefer Option 1 over Option 2, and am strongly against Option 4.

General thoughts:
1. There should only be one way to report metrics.
    i. This alone feels like enough of a reason to me to not introduce Option 4.
2. The action of checkpointing should save a snapshot of data at that exact point in time.
    i. This makes me hesitant about Option 2, as I'm not sure what value it implicitly provides.
3. There should be a common interface for saving and restoring checkpoints (e.g. `Checkpoint`). 
   i. Implicitly converting a directory into a `Checkpoint` is completely unintuitive.
4. APIs should be intuitive.
   i. I actually think the semantics of `session.checkpoint_dir()` are very unclear, can easily be misinterpreted, and may be difficult to maintain over time due to ambiguity of the functionality and requirements.

**Comment by gjoliver** [comment]
can I ask why is it a requirement that we need to have equal number of metrics and checkpoints?
I would like this modified version of proposal 2 a lot more:
```
def train(config):
    if should_checkpoint:
        with session.checkpoint_dir() as checkpoint_dir:
            with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
                f.write("Data\n")

    # Almost every single iteration.
    session.report({"metric": 1})
```

**Comment by krfricke** [comment]
There has been some follow-up discussion offline about this.

For now we are not considering 4, though we may revisit this in the future, pending other discussions about the Checkpoint API.

Option 1 and 2 are identical in terms of reporting checkpoints and metrics. The only difference is that in 1) users create the checkpoint dir and 2) Ray AIR creates the checkpoint dir.

### User creating the checkpoint dir

Flow:
- The user has to create the directory
- The user writes data into the directory
- The user reports the checkpoint
- We snapshot the data
- The user has to clean up the directory (or keep it)

Pros:
- Allows incremental checkpointing: The user can have one dir and just always write the latest state into it. We snapshot it and it remains for the user to use

Cons:
- We _have_ to snapshot the data, i.e. make a full copy. This will at least double disk space requirements
- We can't optimize data transfer for the user (e.g. move instead of copy, provide an auto-mounted NFS directory)
- If we do want to optimize anyway, we'll _implicitly_ take over control of the directory and move it away. This is unexpected behavior and could be unintended
- The user has to clean up. If they don't, they'll leak disk space
- If using dir checkpoints is a common task, we may want to have a utility to create such dirs anyway

### Ray AIR creating the checkpoint dir

Flow:
- The user requests a directory
- The user writes data into the directory
- The user reports the checkpoint
- Ray AIR does whatever it wants. The data is already snapshotted.

Pros:
- Ray AIR creates directories at the final location or moves them there which is efficient
- Ray AIR can further optimize the checkpoint dir interface by mounting the directories e.g. on EFS
- It is explicit that Ray AIR has control over the directory.
- Ray AIR cleans up directories if they are not used anymore/if a trial fails/finishes without reporting, etc

Cons:
- Another session API (but arguably already existed in tune)
- We can't _force_ a user to use the AIR provided directory. They could still pass a different one.

### Proposal (to be discussed)

The proposal is to allow both options, which is in line with the current `Checkpoint.from_directory(any_directory)` API.

However, behavior will differ for user-provided and AIR-provided directories.

- `session.checkpoint_dir()` will create an AIR checkpoint directory.
- This means the directory is fully managed by AIR
- After exiting the context, the directory may be moved, deleted, anything. It shouldn't be accessed by the user
- This is a clear contract for this API
- User-provided checkpoint directories will never be touched. We will always just snapshot them.

The nice thing here is that we explicitly contain the contract in the `session.checkpoint_dir()` API. Once the context is exited, the user does not have to deal with the directory anymore.

Implementation:

- `session.checkpoint_dir()` drops a marker in the directory
- If `session.report(metrics, checkpoint)` is called with a user-provided checkpoint directory, we will always copy (i.e. snapshot it)
- If `session.report(metrics, checkpoint)` is called with a AIR-provided checkpoint directory (with the marker), we can do whatever we want

We'll have another offline discussion about this later and then converge to a solution.

**Comment by krfricke** [comment]
> can I ask why is it a requirement that we need to have equal number of metrics and checkpoints? I would like this modified version of proposal 2 a lot more:
> 
> ```
> def train(config):
>     if should_checkpoint:
>         with session.checkpoint_dir() as checkpoint_dir:
>             with open(os.path.join(checkpoint_dir, "checkpoint.txt"), "w") as f:
>                 f.write("Data\n")
> 
>     # Almost every single iteration.
>     session.report({"metric": 1})
> ```

There is no requirement for this, having a different number of checkpoints is a common use case. All discussed options here should support this.

**Comment by Yard1** [comment]
I would like to propose another option, which would be similar to option 4 but instead of introducing a new API, we'd reuse `session.report`. We can make it so that it can be used both as a function (like it is currently) and a context manager. Therefore, the two ways of calling it would be:

```python
cp = Checkpoint.from_dict({"foo": "bar"})
session.report(metrics, checkpoint=cp)
```

and

```python
with session.report(metrics) as checkpoint_dir:
    model.save(checkpoint_dir)
```

That way there is no ambiguity as to whether metrics are reported or not and we avoid adding a new API, instead extending the current one. This also optimizes for the directory checkpoint case.

If needed, we can also allow users to pass a checkpoint class:

```python
with session.report(metrics, checkpoint_cls=TorchCheckpoint) as checkpoint_dir:
    model.save(checkpoint_dir)
```

And we can use overloads and exceptions to ensure that certain arguments are only allowed in either function or context manager case.

**Comment by krfricke** [comment]
For the context-manager option we will have to figure out how to handle framework-specific checkpoints.

Strawman proposal: We have framework-specific functions to save things we need for prediction into the directory. The rest is up to the user.

Example:

```
from ray.train.util import save_dict, load_dict
from ray.train.torch import TorchCheckpoint, save_torch_model

model = SomeTorchModel()
optimizer = SGD()

with session.report(metrics, checkpoint_cls=TorchCheckpoint) as checkpoint_dir:
    save_torch_model(model, checkpoint_dir)
    save_dict(optimizer.state_dict(), checkpoint_dir)

# ...
checkpoint = result.checkpoint


with checkpoint.as_directory() as checkpoint_dir:
    optimizer_state = load_dict(checkpoint_dir)

# ...

# Will load the torch model
predictor = BatchPredictor.from_checkpoint(checkpoint)  
```

**Comment by Yard1** [comment]
Perhaps we can simply expect files with certain names to be present, and throw exceptions (or at least warnings) if they aren't?
